File: apps/api/engine/scanner.py
import uuid
from datetime import datetime, timezone
from .models import Finding, Severity, ScanResult, ScanSummary, OWASPCategory, SEVERITY_SCORE_WEIGHTS
from .parser import MCPConfig
from .checks.secrets import _is_pinned, _is_package_arg
from .checks import (
    check_secrets,
    check_supply_chain,
    check_tool_poisoning,
    check_privilege,
    check_shadow,
    check_code_execution,
    check_osv,
    check_audit,
    check_lifecycle,
    check_config_level,
)

_SEVERITY_ORDER: dict[Severity, int] = {
    Severity.CRITICAL: 0,
    Severity.HIGH: 1,
    Severity.MEDIUM: 2,
    Severity.LOW: 3,
    Severity.INFO: 4,
}

_AT005_HIGH_SERVER_THRESHOLD = 10


def scan(config: MCPConfig) -> ScanResult:
    all_findings: list[Finding] = []
    # Collect per-server findings so config-level checks can cross-reference them
    per_server: dict[str, list[Finding]] = {}

    for server in config.servers:
        server_findings: list[Finding] = []
        server_findings.extend(check_secrets(server))
        server_findings.extend(check_supply_chain(server))
        server_findings.extend(check_tool_poisoning(server))
        server_findings.extend(check_privilege(server))
        server_findings.extend(check_shadow(server))
        server_findings.extend(check_code_execution(server))
        server_findings.extend(check_audit(server))
        server_findings.extend(check_lifecycle(server))
        # Network check (OSV.dev) — may add latency, fails gracefully if unavailable
        server_findings.extend(check_osv(server))

        per_server[server.name] = server_findings
        all_findings.extend(server_findings)

    # Config-level cross-server checks
    all_findings.extend(check_config_level(config, per_server))

    n = len(config.servers)

    # AT-001: Systemic absence of version pinning across the entire config
    if n >= 2:
        pinned_count = sum(1 for s in config.servers if _any_pinned(s))
        if pinned_count == 0:
            all_findings.append(Finding(
                check_id="AT-001",
                title="No version pinning across any configured server",
                detail=(
                    f"None of the {n} configured MCP server(s) pin their package versions. "
                    "Without version pins, every Claude Desktop or Cursor restart may silently pull a "
                    "different (potentially compromised) version of each server package."
                ),
                severity=Severity.MEDIUM,
                owasp=OWASPCategory.MCP08,
                server_name="(all servers)",
                remediation=(
                    "Pin all package versions in server args "
                    "(e.g., `@modelcontextprotocol/server-filesystem@1.2.3`). "
                    "This ensures reproducibility and protects against silent rug pulls."
                ),
                engine="custom",
            ))

    # AT-005: Excessive server count — each server is an independent entry point
    if n >= _AT005_HIGH_SERVER_THRESHOLD:
        all_findings.append(Finding(
            check_id="AT-005",
            title=f"Excessive MCP server count: {n} servers configured",
            detail=(
                f"This config registers {n} MCP servers. Each server is an independent process "
                "with its own tool set, permissions, and attack surface. "
                f"Configurations with {_AT005_HIGH_SERVER_THRESHOLD}+ servers are statistically "
                "more likely to contain at least one compromised or misconfigured server, "
                "and the total attack surface grows with each addition. "
                "Real-world research found that 70% of MCP servers have at least one security finding."
            ),
            severity=Severity.INFO,
            owasp=OWASPCategory.MCP08,
            server_name="(all servers)",
            remediation=(
                f"Audit each of the {n} servers and remove any that are no longer actively used. "
                "For each server, confirm you understand what permissions it has and why it needs them. "
                "A smaller, well-audited set of servers is safer than a large unreviewed collection."
            ),
            engine="custom",
        ))

    all_findings.sort(key=lambda f: _SEVERITY_ORDER.get(f.severity, 99))

    return ScanResult(
        scan_id=str(uuid.uuid4()),
        config_hash=config.config_hash,
        findings=all_findings,
        summary=_summarize(all_findings, n),
        scanned_at=datetime.now(timezone.utc).isoformat(),
    )


def _any_pinned(server) -> bool:
    """True if the server has at least one package arg pinned to an exact semver."""
    for arg in server.args:
        if _is_package_arg(arg) and _is_pinned(arg):
            return True
    return False


def _risk_score(findings: list[Finding]) -> tuple[int, str]:
    """Return (0-100 risk score, A-F grade)."""
    raw = sum(SEVERITY_SCORE_WEIGHTS.get(f.severity.value, 0) for f in findings)
    score = min(100, raw)
    if score < 20:
        grade = "A"
    elif score < 40:
        grade = "B"
    elif score < 60:
        grade = "C"
    elif score < 80:
        grade = "D"
    else:
        grade = "F"
    return score, grade


def _summarize(findings: list[Finding], server_count: int) -> ScanSummary:
    counts: dict[Severity, int] = {s: 0 for s in Severity}
    owasp_hit: set[str] = set()
    for f in findings:
        counts[f.severity] += 1
        owasp_hit.add(f.owasp.value)
    score, grade = _risk_score(findings)
    return ScanSummary(
        total=len(findings),
        critical=counts[Severity.CRITICAL],
        high=counts[Severity.HIGH],
        medium=counts[Severity.MEDIUM],
        low=counts[Severity.LOW],
        info=counts[Severity.INFO],
        servers_scanned=server_count,
        owasp_coverage=sorted(owasp_hit),
        risk_score=score,
        risk_grade=grade,
    )

File: packages/cli/internal/engine/scanner.go
package engine

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/checks"
	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/models"
	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/parser"
)

// ScanOptions controls optional behaviour of the local engine.
type ScanOptions struct {
	NoNetwork  bool          // skip OSV.dev CVE lookup
	OSVTimeout time.Duration // timeout for OSV queries (default 3s)
}

var severityOrder = map[models.Severity]int{
	models.SeverityCritical: 0,
	models.SeverityHigh:     1,
	models.SeverityMedium:   2,
	models.SeverityLow:      3,
	models.SeverityInfo:     4,
}

const at005HighServerThreshold = 10

// scan runs all checks on the parsed config and returns a ScanResult.
func scan(config *parser.MCPConfig, opts ScanOptions) *models.ScanResult {
	osvTimeout := opts.OSVTimeout
	if osvTimeout == 0 {
		osvTimeout = 3 * time.Second
	}

	allFindings := []models.Finding{}
	perServer := map[string][]models.Finding{}

	activeServers := make([]parser.MCPServer, 0, len(config.Servers))
	for _, s := range config.Servers {
		if !s.Disabled {
			activeServers = append(activeServers, s)
		}
	}

	for _, srv := range activeServers {
		s := srv // capture range var
		sf := []models.Finding{}
		sf = append(sf, checks.CheckSecrets(&s)...)
		sf = append(sf, checks.CheckSupplyChain(&s)...)
		sf = append(sf, checks.CheckToolPoisoning(&s)...)
		sf = append(sf, checks.CheckPrivilege(&s)...)
		sf = append(sf, checks.CheckShadow(&s)...)
		sf = append(sf, checks.CheckCodeExecution(&s)...)
		sf = append(sf, checks.CheckAudit(&s)...)
		sf = append(sf, checks.CheckLifecycle(&s)...)
		if !opts.NoNetwork {
			sf = append(sf, checks.CheckOSV(&s, osvTimeout)...)
		}
		perServer[s.Name] = sf
		allFindings = append(allFindings, sf...)
	}

	// Config-level and cross-server checks
	allFindings = append(allFindings, checks.CheckConfigLevel(config, perServer)...)
	allFindings = append(allFindings, checks.CheckCrossServerChains(config, perServer)...)

	n := len(activeServers)

	// AT-001: No version pinning across any configured server
	if n >= 2 {
		pinnedCount := 0
		for _, srv := range activeServers {
			if anyPinned(&srv) {
				pinnedCount++
			}
		}
		if pinnedCount == 0 {
			allFindings = append(allFindings, models.Finding{
				CheckID:    "AT-001",
				Title:      "No version pinning across any configured server",
				Detail:     fmt.Sprintf("None of the %d configured MCP server(s) pin their package versions. Without version pins, every Claude Desktop or Cursor restart may silently pull a different (potentially compromised) version of each server package.", n),
				Severity:   models.SeverityMedium,
				OWASP:      models.MCP08,
				ServerName: "(all servers)",
				Remediation: "Pin all package versions in server args (e.g., `@modelcontextprotocol/server-filesystem@1.2.3`). This ensures reproducibility and protects against silent rug pulls.",
				Engine:   "custom",
				CWEID:    "CWE-1104",
			})
		}
	}

	// AT-005: Excessive server count
	if n >= at005HighServerThreshold {
		allFindings = append(allFindings, models.Finding{
			CheckID:    "AT-005",
			Title:      fmt.Sprintf("Excessive MCP server count: %d servers configured", n),
			Detail:     fmt.Sprintf("This config registers %d MCP servers. Each server is an independent process with its own tool set, permissions, and attack surface. Configurations with %d+ servers are statistically more likely to contain at least one compromised or misconfigured server, and the total attack surface grows with each addition. Real-world research found that 70%% of MCP servers have at least one security finding.", n, at005HighServerThreshold),
			Severity:   models.SeverityInfo,
			OWASP:      models.MCP08,
			ServerName: "(all servers)",
			Remediation: fmt.Sprintf("Audit each of the %d servers and remove any that are no longer actively used. For each server, confirm you understand what permissions it has and why it needs them. A smaller, well-audited set of servers is safer than a large unreviewed collection.", n),
			Engine:   "custom",
		})
	}

	// AT-006: Docker image with unpinned tag
	for _, srv := range activeServers {
		s := srv
		if cmdBase := checks.CmdBasename(&s); cmdBase == "docker" {
			if image := extractDockerImage(s.Args); image != "" && !dockerImagePinned(image) {
				allFindings = append(allFindings, models.Finding{
					CheckID:    "AT-006",
					Title:      fmt.Sprintf("Docker image without pinned tag: `%s`", image),
					Detail:     fmt.Sprintf("Server `%s` uses Docker image `%s` without a specific version tag or SHA digest. Without pinning, every restart can pull a different image version. An attacker who compromises the image registry or performs a tag-overwrite attack can silently replace the running MCP server code. Using `:latest` or no tag is equivalent to `@*` in npm — never reproducible, never auditable.", s.Name, image),
					Severity:   models.SeverityMedium,
					OWASP:      models.MCP04,
					ServerName: s.Name,
					Remediation: fmt.Sprintf("Pin the image to a specific version tag or SHA digest: `%s:1.2.3` or `%s@sha256:<digest>`. SHA digest pinning is stronger (tags are mutable; digests are not). Re-pin intentionally when you want to upgrade, not automatically.", image, image),
					Engine:       "custom",
					AttackTactic: "initial-access",
					CWEID:        "CWE-1104",
				})
			}
		}
	}

	// Sort by severity
	sort.SliceStable(allFindings, func(i, j int) bool {
		ri := severityOrder[allFindings[i].Severity]
		rj := severityOrder[allFindings[j].Severity]
		return ri < rj
	})

	return &models.ScanResult{
		ScanID:     uuid.New().String(),
		ConfigHash: config.ConfigHash,
		Findings:   allFindings,
		Summary:    summarize(allFindings, n),
		ScannedAt:  time.Now().UTC().Format(time.RFC3339),
	}
}

func anyPinned(server *parser.MCPServer) bool {
	for _, arg := range server.Args {
		if checks.IsPackageArg(arg) && checks.IsPinned(arg) {
			return true
		}
	}
	// Python packages: mcp-server==1.2.3 is pinned
	for _, pkg := range checks.ExtractPackages(server) {
		if pkg.Runtime == "pypi" && strings.Contains(pkg.Name, "==") {
			return true
		}
	}

	return false
}

func extractDockerImage(args []string) string {
	runIdx := -1
	for i, a := range args {
		if strings.ToLower(a) == "run" {
			runIdx = i
			break
		}
	}
	if runIdx < 0 {
		return ""
	}
	skipNextArg := map[string]bool{
		"-e": true, "--env": true, "-v": true, "--volume": true, "--name": true,
		"-p": true, "--publish": true, "--network": true, "--user": true, "-u": true,
		"--entrypoint": true, "--workdir": true, "-w": true, "--label": true,
		"-l": true, "--runtime": true, "--platform": true, "--memory": true,
		"-m": true, "--cpus": true, "--add-host": true, "--dns": true, "--env-file": true,
	}
	i := runIdx + 1
	for i < len(args) {
		arg := args[i]
		if strings.HasPrefix(arg, "-") {
			if skipNextArg[arg] {
				i += 2
			} else {
				i++
			}
		} else {
			return arg
		}
	}
	return ""
}

var dockerFloatingTags = map[string]bool{
	"latest": true, "stable": true, "edge": true, "main": true,
	"master": true, "dev": true, "development": true, "test": true, "prod": true,
}

func dockerImagePinned(image string) bool {
	if strings.Contains(image, "@sha256:") {
		return true
	}
	if !strings.Contains(image, ":") {
		return false
	}
	tag := image[strings.LastIndex(image, ":")+1:]
	return !dockerFloatingTags[tag]
}

func summarize(findings []models.Finding, serverCount int) models.ScanSummary {
	counts := map[models.Severity]int{}
	owaspHit := map[string]bool{}
	for _, f := range findings {
		counts[f.Severity]++
		owaspHit[string(f.OWASP)] = true
	}
	score, grade := riskScore(findings)
	coverage := make([]string, 0, len(owaspHit))
	for k := range owaspHit {
		coverage = append(coverage, k)
	}
	sort.Strings(coverage)
	return models.ScanSummary{
		Total:          len(findings),
		Critical:       counts[models.SeverityCritical],
		High:           counts[models.SeverityHigh],
		Medium:         counts[models.SeverityMedium],
		Low:            counts[models.SeverityLow],
		Info:           counts[models.SeverityInfo],
		ServersScanned: serverCount,
		OWASPCoverage:  coverage,
		RiskScore:      score,
		RiskGrade:      grade,
	}
}

func riskScore(findings []models.Finding) (int, string) {
	raw := 0
	for _, f := range findings {
		raw += models.SeverityScoreWeights[f.Severity]
	}
	if raw > 100 {
		raw = 100
	}
	grade := "A"
	switch {
	case raw >= 80:
		grade = "F"
	case raw >= 60:
		grade = "D"
	case raw >= 40:
		grade = "C"
	case raw >= 20:
		grade = "B"
	}
	return raw, grade
}
